src/vs/workbench/parts/extensions/electron-browser/extensionsActions.ts

```diff
@@ -1797,6 +1815,56 @@ export abstract class AbstractConfigureRecommendedExtensionsAction extends Actio
 			}));
 	}
 
+	protected addRecommendedExtensionToFolder(extensionsFileResource: URI, extensionId: string): TPromise<any> {
+		return this.getOrCreateExtensionsFile(extensionsFileResource)
+			.then(({ content }) => {
+				const jsonContent: IExtensionsConfigContent = json.parse(content) || {};
+				const folderRecommendations = jsonContent.recommendations || [];
+
+				if (folderRecommendations.map(e => e.toLowerCase()).indexOf(extensionId.toLowerCase()) !== -1) {
+					return TPromise.as(null);
+				}
+				folderRecommendations.push(extensionId);
+
+				const folderUnwantedRecommedations = jsonContent.unwantedRecommendations || [];
+				let index = -1;
+				for (let i = 0; i < folderUnwantedRecommedations.length; i++) {
+					if (folderUnwantedRecommedations[i].toLowerCase() === extensionId.toLowerCase()) {
+						index = i;
+						break;
+					}
+				}
+
+				let removeFromUnwantedPromise = TPromise.wrap(null);
+				if (index > -1) {
+					folderUnwantedRecommedations.splice(index, 1);
+					removeFromUnwantedPromise = this.jsonEditingService.write(extensionsFileResource,
+						{
+							key: 'unwantedRecommendations',
+							value: folderUnwantedRecommedations
+						},
+						true);
+				}
+
+				return removeFromUnwantedPromise.then(() =>
+					this.jsonEditingService.write(extensionsFileResource,
+						{
+							key: 'recommendations',
+							value: folderRecommendations
+						},
+						true)
```

Comment:
If the first promise succeeds but the second fails the user would endup with the extension removed from unwanted recommendations but not added to the recommended ones. Not sure if we could safely do anything about it right now because the rollback wouldn't be guaranteed either, and the `jsonEditingService` as it is now apparently can't edit multiple keys in one go. This might be highly unlikely so it's probably fine.

Reply:
This should be fine. The rejected promise would bubble up as an error and shown in the notification.